tools/Environments/DevHome.Environments/ViewModels/PerProviderViewModel.cs

```diff
@@ -0,0 +1,139 @@
+using DevHome.Common.Services;
+using Microsoft.UI.Xaml;
+
+namespace DevHome.Environments.ViewModels;
+
+// View model representing a compute system provider and its associated compute systems.
+public partial class PerProviderViewModel : ObservableObject
+{
+    public string ProviderName { get; }
+
+    public string ProviderID { get; }
+
+    public string DecoratedDevID { get; }
+
+    public ObservableCollection<ComputeSystemCardBase> ComputeSystems { get; }
+
+    public AdvancedCollectionView ComputeSystemAdvancedView { get; set; }
+
+    private readonly StringResource _stringResource;
+
+    private readonly Window _mainWindow;
+
+    [ObservableProperty]
+    private bool _isVisible = true;
+
+    public PerProviderViewModel(string name, string id, string associatedDevID, List<ComputeSystemCardBase> computeSystems, Window mainWindow)
+    {
+        ProviderName = name;
+        ProviderID = id;
+        DecoratedDevID = associatedDevID.Length > 0 ? '(' + associatedDevID + ')' : string.Empty;
+        ComputeSystems = new ObservableCollection<ComputeSystemCardBase>(computeSystems);
+        _mainWindow = mainWindow;
+
+        _stringResource = new StringResource("DevHome.Environments.pri", "DevHome.Environments/Resources");
+        ComputeSystemAdvancedView = new AdvancedCollectionView(ComputeSystems);
+        ComputeSystemAdvancedView.SortDescriptions.Add(new SortDescription("IsCardCreating", SortDirection.Descending));
+    }
+
+    /// <summary>
+    /// Updates the view model to show only the compute systems that match the search criteria.
+    /// </summary>
+    public void SearchHandler(string query)
+    {
+        ComputeSystemAdvancedView.Filter = system =>
+        {
+            if (system is CreateComputeSystemOperationViewModel createComputeSystemOperationViewModel)
+            {
+                return createComputeSystemOperationViewModel.EnvironmentName.Contains(query, StringComparison.OrdinalIgnoreCase);
+            }
+
+            if (system is ComputeSystemViewModel computeSystemViewModel)
+            {
+                var systemName = computeSystemViewModel.ComputeSystem!.DisplayName.Value;
+                var systemAltName = computeSystemViewModel.ComputeSystem.SupplementalDisplayName.Value;
+                return systemName.Contains(query, StringComparison.OrdinalIgnoreCase) || systemAltName.Contains(query, StringComparison.OrdinalIgnoreCase);
+            }
+
+            return false;
+        };
+
+        _mainWindow.DispatcherQueue.EnqueueAsync(() =>
+        {
+            IsVisible = ComputeSystemAdvancedView.Count > 0;
+        });
+    }
+
+    /// <summary>
+    /// Updates the view model to sort the compute systems according to the sort criteria.
+    /// </summary>
+    /// <remarks>
+    /// New SortDescription property names should be added as new properties to <see cref="ComputeSystemCardBase"/>
+    /// </remarks>
+    public void SortHandler(int selectedSortIndex)

```

Comment:
Consider changing sorting to an enumerator.  Dealing with enums is more safe compared to using plain ints.  Additionally, if the sort order changes, the code won't have runtime issues.

Reply:
True, but this value is from the XAML. Don't think I can use an enumerator there.

Reply:
I didn't know you declared all the sorting options in XAML.  A suggestion then, you don't have to do this for this PR, is to use tags.  XAMl can pass the tag of the selected item into the command.